### PR TITLE
Parse commandline flags with gflags.

### PR DESCRIPTION
**Pull request by hcoona**
I'd like to switch google/benchmark to use gflags because my project depends on gflags to control some options. The replacement should be trivial, but some unittests were broken. Need help.

I saw the previous discussion that welcome to switching to gflags. (https://github.com/google/benchmark/issues/320)

### CODE REVIEW

**Comment by hcoona** [comment]
@googlebot I signed it!

**Comment by LebedevRI** [comment]
I'm not sure why this is talking about "switching bazel build to gflags", is cmake build unaffected?

**Comment by dominichamon** [comment]
@LebedevRI only the bazel build has had the dependency added. this is an omission that would need to be fixed before this can be merged.

@hcoona thanks for the PR. I'm happy to get this landed. If you check, you'll see where we're depending on gtest and how we're pulling that in. We'd need something similar in the CMake to find and depend on gflags if you want to support it. That's likely what is failing the tests. Well that, and the CMake still references `commandlineflags_test.cc`.

**Comment by hcoona** [comment]
Thanks for figuring out how to involve external libraries for cmake. I'll fix it a few days later. Quite busy recently.

**Comment by hcoona** [comment]
Is there any CMake experts who might like to give me a hand on the CMake build chain. I followed the steps downloading googletest repository to download gflags. But it failed with the error:

```
CMake Error at cmake/GFlags.cmake:38 (set_target_properties):
  set_target_properties can not be used on an ALIAS target.
Call Stack (most recent call first):
  CMakeLists.txt:266 (include)
```

I think it could because the [CMakeLists.txt file in gflags repository](https://github.com/gflags/gflags/blob/f7388c6/CMakeLists.txt#L507) used the `ALIAS` keyword, but I don't think it reasonable to remove it from gflags.

**Comment by hcoona** [comment]
I tried my best but still failed to make the CMake chain work. I already link the `benchmark` target with `gflags_static`, but the CMake process still failed with the following error:

```
CMake Error: install(EXPORT "benchmarkTargets" ...) includes target "benchmark" which requires target "gflags_static" that is not in the export set.
```

**Comment by dominichamon** [comment]
it looks like you've managed to merge a few of the upstream changes into your change. could you rebase so the PR only contains your changes?

File: CMakeLists.txt [hunk 1]
```diff
@@ -262,6 +262,9 @@ cxx_feature_check(STEADY_CLOCK)
 set(THREADS_PREFER_PTHREAD_FLAG ON)
 find_package(Threads REQUIRED)
 
+# Include 3rd party module gflags
+include(GFlags)
```
Comment (LebedevRI) [inline CMakeLists.txt:266]: Thank you for bothering writing *this* logic, i would have insisted on it if it wasn't there,
but by default it should not be doing that, but relying on pre-installed version,
that should be looked for via traditional `find_package(GFlags)`.

File: cmake/GFlags.cmake.in [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    SOURCE_DIR        "${GFLAGS_PATH}" # use existing src dir.
+    BINARY_DIR        "${CMAKE_BINARY_DIR}/build"
+    CONFIGURE_COMMAND ""
+    BUILD_COMMAND     ""
+    INSTALL_COMMAND   ""
+    TEST_COMMAND      ""
+  )
+else()
+  if(NOT ALLOW_DOWNLOADING_GFLAGS)
+    message(SEND_ERROR "Did not find GFlags sources! Either pass correct path in GFLAGS_PATH, or enable BENCHMARK_DOWNLOAD_DEPENDENCIES, or disable BENCHMARK_ENABLE_GTEST_TESTS / BENCHMARK_ENABLE_TESTING.")
```
Comment (dominichamon) [inline cmake/GFlags.cmake.in:32]: this bit of the error that talks about disabling testing doesn't apply to gflags.

**Comment by dominichamon** [comment]
I'd really like to see this one completed and stop supporting hand-rolled command line logic.

**Pull request closed without merging**